### PR TITLE
IGNITE-9120 Metadata writer does not propagate error to failure handler

### PR DESCRIPTION
**Pull request by a-polyakov**
Signed-off-by: a-polyakov <polyakov.alexandr.alexandrovich@gmail.com>

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/binary/BinaryMetadataFileStore.java [hunk 1]
```diff
@@ -91,17 +94,27 @@ void writeMetadata(BinaryMetadata binMeta) {
...
 
-                out.write(marshalled);
+            if (ctx.config() != null
+                && ctx.config().getDataStorageConfiguration() != null
+                && ctx.config().getDataStorageConfiguration().getFileIOFactory() != null) {
+                try (final FileIO out = ctx.config().getDataStorageConfiguration().getFileIOFactory().create(file)) {
+                    out.write(marshalled, 0, marshalled.length);
+                }
+            }
+            else {
```
Comment (zstan) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/binary/BinaryMetadataFileStore.java]: i suppose no "else" branching needs here

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/IgnitePdsNoSpaceLeftOnDeviceTest.java [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+    /** {@inheritDoc} */
+    @Override protected IgniteConfiguration getConfiguration(String gridName) throws Exception {
+        IgniteConfiguration cfg = super.getConfiguration(gridName);
+
+        final DataStorageConfiguration dataStorageConfiguration = new DataStorageConfiguration();
+
+        dataStorageConfiguration.getDefaultDataRegionConfiguration().setPersistenceEnabled(true);
+        dataStorageConfiguration.setFileIOFactory(new FailingFileIOFactory());
+
+        cfg.setDataStorageConfiguration(dataStorageConfiguration);
```
Comment (zstan) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/IgnitePdsNoSpaceLeftOnDeviceTest.java]: need to set dataRegion size.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/IgnitePdsNoSpaceLeftOnDeviceTest.java [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+        cleanPersistenceDir();
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void afterTest() throws Exception {
+        cleanPersistenceDir();
+    }
+
+    /**
+     * https://issues.apache.org/jira/browse/IGNITE-9120
```
Comment (zstan) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/IgnitePdsNoSpaceLeftOnDeviceTest.java]: do we need ticket link here?

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/IgnitePdsNoSpaceLeftOnDeviceTest.java [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+        private static final long serialVersionUID = 0L;
+
+        /**
+         *
+         */
+        private final FileIOFactory delegateFactory = new RandomAccessFileIOFactory();
+
+        /** {@inheritDoc} */
+        @Override public FileIO create(File file, OpenOption... modes) throws IOException {
+            if (file.getAbsolutePath().contains("node01-") && file.getAbsolutePath().contains("binary_meta"))
```
Comment (zstan) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/IgnitePdsNoSpaceLeftOnDeviceTest.java]: can you replace hardcoded "node01-" here ?

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/IgnitePdsNoSpaceLeftOnDeviceTest.java (hunk 2 above)
Comment (a-polyakov) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/IgnitePdsNoSpaceLeftOnDeviceTest.java]: 1<<24 = 16mb

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/IgnitePdsNoSpaceLeftOnDeviceTest.java (hunk 3 above)
Comment (a-polyakov) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/IgnitePdsNoSpaceLeftOnDeviceTest.java]: Search by "https://issues.apache.org/jira/browse/IGNITE-" issued 527 matches in test

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/binary/BinaryMetadataFileStore.java [hunk 5]
```diff
@@ -91,17 +103,34 @@ void writeMetadata(BinaryMetadata binMeta) {
...
 
         try {
-            File file = new File(workDir, Integer.toString(binMeta.typeId()) + ".bin");
+            File file = new File(workDir, binMeta.typeId() + ".bin");
+
+            byte[] marshalled = U.marshal(ctx, binMeta);
 
-            try(FileOutputStream out = new FileOutputStream(file, false)) {
-                byte[] marshalled = U.marshal(ctx, binMeta);
+            if (fileIOFactory != null) {
```
[the remark this answers is not in the document]
Comment (a-polyakov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/binary/BinaryMetadataFileStore.java]: check fileIOFactory missing in isPersistenceEnabled

**Pull request closed without merging**